src/visitors/sympy_replace_solutions_visitor.cpp

```diff
@@ -0,0 +1,354 @@
+/*************************************************************************
+ * Copyright (C) 2018-2019 Blue Brain Project
+ *
+ * This file is part of NMODL distributed under the terms of the GNU
+ * Lesser General Public License. See top-level LICENSE file for details.
+ *************************************************************************/
+
+#include "visitors/sympy_replace_solutions_visitor.hpp"
+#include "visitors/lookup_visitor.hpp"
+
+
+#include "ast/all.hpp"
+#include "utils/logger.hpp"
+#include "visitors/visitor_utils.hpp"
+
+namespace nmodl {
+namespace visitor {
+
+
+void SympyReplaceSolutionsVisitor::InterleavesCounter::new_equation(const bool is_in_system) {
+    n_ += (!in_system_ && is_in_system);
+    in_system_ = is_in_system;
+}
+
+SympyReplaceSolutionsVisitor::SympyReplaceSolutionsVisitor(
+    const std::vector<std::string>& pre_solve_statements,
+    const std::vector<std::string>& solutions,
+    const std::unordered_set<ast::Statement*>& to_be_removed)
+    : pre_solve_statements_(pre_solve_statements.begin(), pre_solve_statements.end())
+    , to_be_removed_(&to_be_removed) {
+    const auto ss_tmp_delimeter =
+        std::find_if(solutions.begin(), solutions.end(), [](const std::string& statement) {
+            return statement.substr(0, 3) != "tmp";
+        });
+    tmp_statements_ = SolutionSorter(solutions.begin(), ss_tmp_delimeter);
+    solutions_ = SolutionSorter(ss_tmp_delimeter, solutions.end());
+
+    replacements_.clear();
+    is_statement_block_root_ = true;
+}
+
+
+void SympyReplaceSolutionsVisitor::visit_statement_block(ast::StatementBlock& node) {
+    const bool is_root = is_statement_block_root_;
```

Comment:
no. It assumes that the first node where you call it is the root node. In the root node I need to finally replace stuff, set/check 
the interleaves_counter etc. Since I possibly run the visitor twice from the root node I cannot just set everything in the constructor.

Reply:
no, we can call it also on the child nodes. However, in the root node I need to set and update a few things that are global. Since I may need to pass twice I sometimes need to reset stuff (but only if I am the root node)

Reply:
for example: interleaves_counter

Reply:
Yeah, I think we agree. I understood that it is called on the root nodes but not only. The question was more, that necessarily it has to be called first on the root node, that's the entry point. But as I said, this is probably all irrelevant, because it's an internal visitor and we have probably plenty of other visitors that are exactly the same. The assumption is that whoever calls this visitor they need to first call it on the root node, and it will then recursively call on child nodes. I'm going to resolve this, since I don't think it's an actual issue, just wanted to clarify.